File: grind75/queue_using_stack.py
class MyQueue:

    def __init__(self):
        self.data = []
        self.pop_data = []
        
    def push(self, x: int) -> None:
        self.data.append(x)
    def pop(self) -> int:
        while self.data:
            self.pop_data.append(self.data.pop())
        temp = self.pop_data.pop()
        while self.pop_data:
            self.data.append(self.pop_data.pop())
        return temp
    def peek(self) -> int:
        while self.data:
            self.pop_data.append(self.data.pop())
        temp = self.pop_data[-1]
        while self.pop_data:
            self.data.append(self.pop_data.pop())
        return temp
    def empty(self) -> bool:
        if len(self.data) == 0:
            return True
        return False

File: grind75/queue_using_stack.cpp
// 0ms solution

class MyQueue {
private:
    stack<int> data;
    stack<int> temp;
    void interchange(stack<int> *x1, stack<int> *x2){
        while(!x1->empty()){
            cout << x1->top() << endl;
            x2->push(x1->top());
            x1->pop();
        }
    }
public:
    MyQueue() {
    }
    
    void push(int x) {
        interchange(&data, &temp);
        data.push(x);
        interchange(&temp, &data);
    }
    
    int pop() {
        int res = data.top();
        data.pop();
        return res;
    }
    
    int peek() {
        return data.top();
    }
    
    bool empty() {
        return data.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */